src/main/java/co/cask/common/AvroToStructuredTransformer.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.common;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.format.UnexpectedFormatException;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.hydrator.common.RecordConverter;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import org.apache.avro.generic.GenericRecord;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Create StructuredRecords from GenericRecords
+ */
+public class AvroToStructuredTransformer extends RecordConverter<GenericRecord, StructuredRecord> {
+  private final Map<Integer, Schema> schemaCache = Maps.newHashMap();
+
+  public StructuredRecord transform(GenericRecord genericRecord) throws IOException {
+    org.apache.avro.Schema genericRecordSchema = genericRecord.getSchema();
+    return transform(genericRecord, convertSchema(genericRecordSchema));
+  }
+
+  @Override
+  public StructuredRecord transform(GenericRecord genericRecord, Schema structuredSchema) throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(structuredSchema);
+    for (Schema.Field field : structuredSchema.getFields()) {
+      String fieldName = field.getName();
+      Object value = convertField(genericRecord.get(fieldName), field.getSchema());
+      if (value instanceof LocalDate) {
```

Comment:
instanceof is usually a sign that logic is happening in the wrong place. The convertField() method should take care of returning the correct physical type and you should just set whatever value you get back.

Reply:
made corresponding changes.